templates/scripts/yamlSolver.js

```diff
@@ -0,0 +1,132 @@
+const { writeFileSync, readFileSync, lstatSync, existsSync } = require("node:fs");
+const path = require("path");
+const utils = require("./utils");
+
+// The solver is called by the following command:
+// > node yamlSolver.js <action> <constraintFilePath>
+
+// The solver support two actions
+// 1. apply the constraints to solution
+// 2. verify the constraints are satisfied.
+const Action = {
+  APPLY: "apply",
+  VERIFY: "verify",
+};
+
+// The constraints are defined in mustache files.
+// default constraints folders
+const mustacheFolder = path.resolve(__dirname, "..", "constraints", "yml", "templates");
+const solutionFolder = path.resolve(__dirname, "..");
+
+// example:  " key1: value, key2 " => { key1: value, key2: true }
+function strToObj(str) {
+  const properties = str.split(",");
+  let obj = {};
+  properties.forEach(function (property) {
+    if (property.includes(":")) {
+      const tup = property.split(":");
+      obj[tup[0].trim()] = tup[1].trim();
+    } else {
+      obj[property.trim()] = true;
+    }
+  });
+  return obj;
+}
+
+// read all yml files and mustache files in folder as mustache variables
+function generateVariablesFromSnippets(dir) {
+  let result = {};
+  utils.filterYmlFiles(dir).map((file) => {
+    const yml = readFileSync(file, "utf8");
+    result = { ...result, ...{ [path.basename(file, ".yml")]: yml } };
+  });
+  utils.filterMustacheFiles(dir).map((file) => {
+    const mustache = readFileSync(file, "utf8");
+    result = {
+      ...result,
+      ...{
+        [path.basename(file, ".mustache")]: function () {
+          return function (text) {
+            return utils.renderMustache(mustache, strToObj(text)).trimEnd();
+          };
+        },
+      },
+    };
+  });
+  return result;
+}
+
+function* solveMustache(mustachePaths) {
```

Comment:
What's the special purpose of adding * after function key word?

Reply:
function* declares it returns an iterator. I may refactor this function to make the code style more consistent later.